### PR TITLE
Fix trait solving ICEs

### LINKED ISSUE #77653: [ICE] Encountered errors resolving bounds after type-checking
The below content was posted by @bkchr, I moved it into a separated issue to keep track of the ICE.

---

Hey, so I tried porting our code to compile with latest nightly and I almost got it working but ultimately it ended with an internal compiler error:
```
error: internal compiler error: compiler/rustc_trait_selection/src/traits/codegen/mod.rs:121:9: Encountered errors `[FulfillmentError(Obligation(predicate=TraitPredicate(<dyn sc_client_api::PrunableStateChangesTrieStorage<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, sp_runtime::traits::BlakeTwo256>, sp_runtime::OpaqueExtrinsic>> as sp_state_machine::changes_trie::Storage<sp_runtime::traits::BlakeTwo256, u32>>), depth=0),Unimplemented)]` resolving bounds after type-checking

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:945:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.49.0-nightly (98edd1fbf 2020-10-06) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: aborting due to previous error

error: could not compile `node-testing`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

I created a branch that fails with this or a similar ICE when you run `cargo test --all`. You can find it here: https://github.com/paritytech/substrate/tree/bkchr-fix-for-latest-nightly

_Originally posted by @bkchr in https://github.com/rust-lang/rust/issues/77638#issuecomment-704892157_

@rustbot modify labels:  regression-from-stable-to-nightly E-needs-mcve A-associated-items

### LINKED ISSUE #77656: ICE: Projection bound ... was applicable to ... but now is not
I noticed that my crate https://github.com/tavianator/acap doesn't build with the latest nightly.  Here's a reduced testcase:

### Code

```Rust
pub trait Value: PartialOrd {}

impl<T: PartialOrd> Value for T {}

pub trait Distance
where
    Self: PartialOrd<<Self as Distance>::Value>,
    Self: PartialOrd,
{
    type Value: Value;
}

impl<T: Value> Distance for T {
    type Value = T;
}

pub trait Proximity<T = Self> {
    type Distance: Distance;
}

pub type DistanceValue<K, V = K> = <<K as Proximity<V>>::Distance as Distance>::Value;

struct VpNode<T, R = DistanceValue<T>> {
    item: T,
    radius: R,
}
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0085a0f26ba32eace4c638a3663606a7

### Meta

`rustc --version --verbose`:
```
rustc 1.49.0-nightly (98edd1fbf 2020-10-06)
binary: rustc
commit-hash: 98edd1fbf8a68977a2a7c1312eb1ebff80515a92
commit-date: 2020-10-06
host: x86_64-unknown-linux-gnu
release: 1.49.0-nightly
LLVM version: 11.0
```

### Error output

```
error: internal compiler error: compiler/rustc_trait_selection/src/traits/select/confirmation.rs:158:25: Projection bound `Binder(<<Self as Proximity<T>>::Distance as std::cmp::PartialOrd<<<Self as Proximity<T>>::Distance as Distance>::Value>>)` was applicable to `Obligation(predicate=Binder(TraitPredicate(<<Self as Proximity<T>>::Distance as std::cmp::PartialOrd>)), depth=0)` but now is not
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_trait_selection/src/traits/select/confirmation.rs:158:25: Projection bound `Binder(<<Self as Proximity<T>>::Distance as std::cmp::PartialOrd<<<Self as Proximity<T>>::Distance as Distance>::Value>>)` was applicable to `Obligation(predicate=Binder(TraitPredicate(<<Self as Proximity<T>>::Distance as std::cmp::PartialOrd>)), depth=0)` but now is not

thread 'rustc' panicked at 'Box<Any>', compiler/rustc_errors/src/lib.rs:945:9
stack backtrace:
   0: std::panicking::begin_panic
   1: rustc_errors::HandlerInner::bug
   2: rustc_errors::Handler::bug
   3: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
   4: rustc_middle::ty::context::tls::with_opt::{{closure}}
   5: rustc_middle::ty::context::tls::with_opt
   6: rustc_middle::util::bug::opt_span_bug_fmt
   7: rustc_middle::util::bug::bug_fmt
   8: rustc_trait_selection::traits::select::confirmation::<impl rustc_trait_selection::traits::select::SelectionContext>::confirm_projection_candidate::{{closure}}::{{closure}}
   9: rustc_infer::infer::InferCtxt::commit_unconditionally
  10: rustc_trait_selection::traits::select::confirmation::<impl rustc_trait_selection::traits::select::SelectionContext>::confirm_candidate
  11: rustc_trait_selection::traits::select::SelectionContext::select
  12: rustc_trait_selection::traits::fulfill::FulfillProcessor::process_trait_obligation
  13: rustc_trait_selection::traits::fulfill::FulfillProcessor::progress_changed_obligations
  14: rustc_data_structures::obligation_forest::ObligationForest<O>::process_obligations
  15: <rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::select_where_possible
  16: <rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::select_all_or_error
  17: rustc_typeck::check::fn_ctxt::FnCtxt::select_all_obligations_or_error
  18: rustc_infer::infer::InferCtxtBuilder::enter
  19: rustc_typeck::check::wfcheck::check_associated_item
  20: rustc_typeck::check::wfcheck::check_trait_item
  21: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::check_trait_item_well_formed>::compute
  22: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  23: rustc_data_structures::stack::ensure_sufficient_stack
  24: rustc_query_system::query::plumbing::get_query_impl
  25: rustc_query_system::query::plumbing::ensure_query_impl
  26: rustc_data_structures::sync::par_for_each_in
  27: rustc_typeck::check::check::check_wf_new
  28: rustc_typeck::check_crate
  29: rustc_interface::passes::analysis
  30: rustc_middle::ty::query::<impl rustc_query_system::query::config::QueryAccessors<rustc_middle::ty::context::TyCtxt> for rustc_middle::ty::query::queries::analysis>::compute
  31: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  32: rustc_data_structures::stack::ensure_sufficient_stack
  33: rustc_query_system::query::plumbing::get_query_impl
  34: rustc_interface::passes::QueryContext::enter
  35: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  36: rustc_span::with_source_map
  37: rustc_interface::interface::create_compiler_and_run
  38: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.49.0-nightly (98edd1fbf 2020-10-06) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_trait_item_well_formed] checking that `Proximity::Distance` is well-formed
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error
```

</p>
</details>

### PR DESCRIPTION
**Pull request by matthewjasper**
- Selection candidates that are known to be applicable are preferred
  over candidates that are not.
- Don't ICE if a projection/object candidate is no longer applicable
  (this can happen due to cycles in normalization)
- Normalize supertraits when finding trait object candidates

Closes #77653
Closes #77656

r? @nikomatsakis

### CODE REVIEW

**Comment by tavianator** [comment]
I can confirm that this fixes https://github.com/rust-lang/rust/issues/77656, and https://github.com/tavianator/acap builds successfully now.  Thanks!

**Comment by spastorino** [comment]
@bors p=1

When this is merged it needs high priority because it fixes  `P-critical` issues.

**Comment by lqd** [comment]
If this needs "high priority", over rollups, it needs at least p=6: otherwise priorities 1 to 5 can be skipped by rollups

**Comment by lqd** [comment]
I wasn't aware of that policy either before Manish mentioned it to me.

then let's do @bors p=6 for when the PR is reviewed

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 1]
```diff
@@ -1384,9 +1384,11 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             }
 
             (ProjectionCandidate(i), ProjectionCandidate(j)) => {
-                // Arbitrarily pick the first candidate for backwards
+                // Arbitrarily pick the lower numbered candidate for backwards
                 // compatibility reasons. Don't let this affect inference.
-                i > j && !needs_infer
+                other.evaluation.must_apply_modulo_regions()
+                    && !needs_infer
+                    && (i < j || !victim.evaluation.must_apply_modulo_regions())
```
Comment (nikomatsakis) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1392]: Hmm I was worried when I read the PR description, but this is related narrowly tailored

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
So, r=me, but I'd be happy to talk about this more with @matthewjasper -- they seemed to indicate a desire to talk it over. I'm like a *little* nervous about the changes to prefer one predicate over another but these do seem pretty narrowly tailored, and they avoid inference effects, so I'm not that worried.

**Comment by matthewjasper** [comment]
After trying to find other options here there doesn't seem to be a much better solution without a much larger change here (e.g. making rustc more like chalk). I think that I slightly prefer this change, which simplifies how winnowing works in exchange for evaluating normalization cycles to `Recur` instead of `Unknown`. The patch here is against the head of this PR.

```patch
@@ -521,12 +521,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                             result
                         }
                         Ok(Ok(None)) => Ok(EvaluatedToAmbig),
-                        // EvaluatedToRecur might also be acceptable here, but use
-                        // Unknown for now because it means that we won't dismiss a
-                        // selection candidate solely because it has a projection
-                        // cycle. This is closest to the previous behavior of
-                        // immediately erroring.
-                        Ok(Err(project::InProgress)) => Ok(EvaluatedToUnknown),
+                        Ok(Err(project::InProgress)) => Ok(EvaluatedToRecur),
                         Err(_) => Ok(EvaluatedToErr),
                     }
                 }
@@ -1387,9 +1382,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             | (ObjectCandidate(i), ObjectCandidate(j)) => {
                 // Arbitrarily pick the lower numbered candidate for backwards
                 // compatibility reasons. Don't let this affect inference.
-                other.evaluation.must_apply_modulo_regions()
-                    && !needs_infer
-                    && (i < j || !victim.evaluation.must_apply_modulo_regions())
+                i < j && !needs_infer
             }
             (ObjectCandidate(_), ProjectionCandidate(_))
             | (ProjectionCandidate(_), ObjectCandidate(_)) => {
```

**Comment by nikomatsakis** [comment]
@matthewjasper hmm I think I might prefer that patch too. `EvaluateToRecur` does seem appropriate for a cycle involving normalization.

**Comment by JohnTitor** [comment]
@bors r-
Let me tell bors that this should be tested.

**Comment by JohnTitor** [comment]
So, I'd @bors r- until we figure out why the test isn't triggered. Related Zulip topic: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/bors.20won't.20test.20.2377720

**Comment by JohnTitor** [comment]
Pietro found the cause, re-adding this to the queue.
@bors r=nikomatsakis

**Comment by Aaron1011** [comment]
Bors is stuck

@bors retry

**Comment by matthewjasper** [comment]
seeing if a force-push helps
@bors r=nikomatsakis

**Comment by matthewjasper** [comment]
Hmm, merge was against 8f0fa9d51ff4ad2c0869e660856cd327e79915e9 but master is at 6b9fbf212a06944ff23325d2d63215805dc3c6c3 cc @rust-lang/infra

**Pull request merged**